### ISSUE TITLE
Application of settings doesn't seem to be atomic, allowing failed document loads

### ISSUE THREAD

**Issue by aidanhs** [meilisearch/MeiliSearch#1159]
**Describe the bug**
Updating settings on an index and then immediately attempting to load a document (without waiting for the update to complete) can cause spurious failures relating to invalid attributes for faceting.

I hypothesise this is caused by not having atomicity in settings application (i.e. documents can be loaded while settings are partially applied).

**To Reproduce**
Use my patch (listed at the bottom) to build and run MeiliSearch in release mode (`rm -rf data.ms/* && (cd MeiliSearch/ && cargo build --release) && RUST_BACKTRACE=1 ./MeiliSearch/target/release/meilisearch --db-path data.ms`).
Run the test script (listed at the bottom) against this fresh MeiliSearch instance.

The script logs will look like:
```
$ python3 repro.py
Deleting index
Making index
Response:  b'{"name":"songs","uid":"songs","createdAt":"2021-01-01T14:36:45.046881532Z","updatedAt":"2021-01-01T14:36:45.046929438Z","primaryKey":null}'
Configuring index
Response:  b'{"updateId":0}'
Beginning song load
Songs loading with update id 1
{'status': 'failed', 'updateId': 1, 'type': {'name': 'DocumentsAddition', 'number': 1}, 'error': 'error processing facet filter: invalid document attribute 640, accepted types: String and [String]', 'errorType': 'invalid_request_error', 'errorCode': 'invalid_facet', 'errorLink': 'https://docs.meilisearch.com/errors#invalid_facet', 'duration': 0.0045505, 'enqueuedAt': '2021-01-01T14:36:45.067951958Z', 'processedAt': '2021-01-01T14:36:45.075497909Z'}
Traceback (most recent call last):
  File "repro.py", line 62, in <module>
    updatesearch()
  File "repro.py", line 58, in updatesearch
    raise Exception('failed to process song load')
Exception: failed to process song load
```

The MeiliSearch logs will look like:
```
[2021-01-01T14:36:45Z INFO  actix_web::middleware::logger] 127.0.0.1:40454 "GET /indexes/songs/updates/1 HTTP/1.1" 200 127 "-" "Python-urllib/3.6" 0.000168
attr song_key => FieldId(0) (value "1")
attr title => FieldId(1) (value "succducc - me & u")
attr song_name => FieldId(2) (value "me & u")
attr song_sub_name => FieldId(3) (value "")
attr level_author_name => FieldId(4) (value "datkami")
attr description => FieldId(5) (value "Hard Only / ~330 notes / Event Lighting! / https://soundcloud.com/succducc/me-n-u")
attr total_votes => FieldId(6) (value 640)
attr pct_upvoted => FieldId(7) (value 82.65625)
attr uploaded_at_tstamp => FieldId(8) (value 1525786136)
attr post_id => FieldId(9) (value 7809)
attr uploader_name => FieldId(10) (value "datkami")
attr genres => FieldId(11) (value ["Anime","EDM"])
attr modes => FieldId(12) (value ["Standard"])
attributes for faceting: SetBuf([FieldId(5), FieldId(6), FieldId(7)])
considering document id DocumentId(0)
found field id FieldId(5) with value "\"Hard Only / ~330 notes / Event Lighting! / https://soundcloud.com/succducc/me-n-u\""
found field id FieldId(6) with value "640"
```

As a guide to these latter logs:
1. for the single document being inserted, it prints out each field and the corresponding field ID (as well as the value)
2. it then logs which attribute field ids are going to be faceted
3. for each attribute being faceted, it then logs the value

The output from step 2 is odd - it seems to say that fields 5, 6 and 7 are for faceting, which corresponds to description, total_votes and pct_upvoted. But as you can see from the test script, that's totally wrong. It's almost as if the field IDs have changed behind the scenes, but the fields for faceting haven't.

**Expected behavior**
My main expectation here is atomicity in settings update (i.e. when a set of documents are loaded, they are either loaded wholly under the old settings or new settings - never an inbetween state).

I am familiar with the [asynchronous update behaviour](https://docs.meilisearch.com/guides/advanced_guides/asynchronous_updates.html) of MeiliSearch. I believe this atomicity should hold even if the settings update is not complete - otherwise API consumers need to cooperatively conduct locking.

**Desktop (please complete the following information):**
 - OS: Linux
 - Browser N/A
 - Version Ubuntu 18.04

**Additional context**
Here's the patch, applied against a05aef5c14d445139addb5e8709acaf427cec0cd:
```
diff --git a/meilisearch-core/src/facets.rs b/meilisearch-core/src/facets.rs
index c4689ee..ac0f1a8 100644
--- a/meilisearch-core/src/facets.rs
+++ b/meilisearch-core/src/facets.rs
@@ -190,12 +190,14 @@ pub fn facet_map_from_docids(
     // string with it's case preserved, and a list of document ids for that facet attribute.
     let mut facet_map: HashMap<FacetKey, (String, Vec<DocumentId>)> = HashMap::new();
     for document_id in document_ids {
+        println!("considering document id {:?}", document_id);
         for result in index
             .documents_fields
             .document_fields(rtxn, *document_id)?
         {
             let (field_id, bytes) = result?;
             if attributes_for_facetting.contains(&field_id) {
+                println!("found field id {:?} with value {:?}", field_id, String::from_utf8_lossy(bytes));
                 match serde_json::from_slice(bytes)? {
                     Value::Array(values) => {
                         for v in values {
diff --git a/meilisearch-core/src/settings.rs b/meilisearch-core/src/settings.rs
index f26865d..8627ab6 100644
--- a/meilisearch-core/src/settings.rs
+++ b/meilisearch-core/src/settings.rs
@@ -43,6 +43,7 @@ fn deserialize_some<'de, T, D>(deserializer: D) -> Result<Option<T>, D::Error>
 impl Settings {
     pub fn to_update(&self) -> Result<SettingsUpdate, RankingRuleConversionError> {
         let settings = self.clone();
+        println!("{:?}", settings);
 
         let ranking_rules = match settings.ranking_rules {
             Some(Some(rules)) => UpdateState::Update(RankingRule::try_from_iter(rules.iter())?),
diff --git a/meilisearch-core/src/update/documents_addition.rs b/meilisearch-core/src/update/documents_addition.rs
index b783ae9..53739d9 100644
--- a/meilisearch-core/src/update/documents_addition.rs
+++ b/meilisearch-core/src/update/documents_addition.rs
@@ -229,6 +229,7 @@ pub fn apply_addition(
         // For each key-value pair in the document.
         for (attribute, value) in document {
             let (field_id, _) = schema.insert_with_position(&attribute)?;
+            println!("attr {} => {:?} (value {})", attribute, field_id, value);
             index_document(
                 writer,
                 index.documents_fields,
@@ -260,6 +261,7 @@ pub fn apply_addition(
 
     // recompute all facet attributes after document update.
     if let Some(attributes_for_facetting) = index.main.attributes_for_faceting(writer)? {
+        println!("attributes for faceting: {:?}", attributes_for_facetting);
         let docids = index.main.internal_docids(writer)?;
         let facet_map = facets::facet_map_from_docids(writer, index, &docids, attributes_for_facetting.as_ref())?;
         index.facets.add(writer, facet_map)?;
```

Here's the test script:
```
#!/usr/bin/env python3

import datetime
import json
from urllib import request
import time
import sys

assert sys.version_info.major == 3

def updatesearch():
    print("Deleting index")
    r = request.Request('http://localhost:7700/indexes/songs', method='DELETE')
    try:
        request.urlopen(r)
    except Exception as e:
        if e.getcode() != 404:
            raise

    songs = [{
        'song_key': '1', 'title': 'succducc - me & u', 'song_name': 'me & u', 'song_sub_name': '', 'level_author_name': 'datkami', 'description': 'Hard Only / ~330 notes / Event Lighting! / https://soundcloud.com/succducc/me-n-u', 'total_votes': 640, 'pct_upvoted': 82.65625, 'uploaded_at_tstamp': 1525786136, 'post_id': 7809, 'uploader_name': 'datkami', 'genres': ['Anime', 'EDM'], 'modes': ['Standard']
    }]

    print("Making index")
    r = request.Request('http://localhost:7700/indexes', method='POST',
            data=b'{"uid":"songs"}', headers={'Content-Type': 'application/json'})
    res = request.urlopen(r)
    print('Response: ', res.read())

    print("Configuring index")
    settings = {
        "searchableAttributes": ["title", "song_name", "song_sub_name", "level_author_name", "description", "uploader_name", "genres", "modes"],
        "attributesForFaceting": ["uploader_name", "genres", "modes"],
        "displayedAttributes": ["*"]
    }
    settings = json.dumps(settings).encode('utf8')
    r = request.Request('http://localhost:7700/indexes/songs/settings', method='POST',
            data=settings, headers={'Content-Type': 'application/json'})
    res = request.urlopen(r)
    print('Response: ', res.read())

    print("Beginning song load")
    r = request.Request('http://localhost:7700/indexes/songs/documents?primaryKey=song_key', method='POST',
            data=json.dumps(songs[:1]).encode('utf8'), headers={'Content-Type': 'application/json'})
    res = request.urlopen(r)
    update_id = json.loads(res.read())['updateId']
    print('Songs loading with update id', update_id)
    while True:
        r = request.Request('http://localhost:7700/indexes/songs/updates/'+str(update_id))
        res = request.urlopen(r)
        data = json.loads(res.read())
        if data['status'] == 'enqueued':
            time.sleep(1)
        elif data['status'] == 'processed':
            break
        else:
            print(data)
            raise Exception('failed to process song load')

    print("Song indexing complete")

updatesearch()
```

**Comment by Kerollmops** [comment]
Hey @aidanhs,

Thank you for the report this should be fixed by #1176, I will close this issue but feel free to reopen it!

Have a good day!

**Comment by aidanhs** [comment]
Can confirm this seems to be fixed in 0.19.0
